src/main.ts

```diff
@@ -904,14 +914,18 @@ function generateEncode(ctx: Context, fullName: string, messageDesc: DescriptorP
             `
           : writeSnippet(`{ ${maybeTypeField} key: key as any, value }`);
         chunks.push(code`
-          Object.entries(message.${fieldName}).forEach(([key, value]) => {
-            ${entryWriteSnippet}
-          });
+          if (message.${fieldName} !== undefined && Object.keys(message.${fieldName}).length !== 0) {
+            Object.entries(message.${fieldName}).forEach(([key, value]) => {
```

Comment:
I would if we could make this `Object.entries(message.foo || []).forEach`. I know it's slightly less efficient, in terms of making a throw away `[]` and `lambda` if the field is not set, but I'm thinking that @webmaster128 just did several PRs that a lot of explicit `message.field !== undefined && ...` checks to reduce the code size ouptut.

So, I suppose it's a trade-off, do you prefer less code w/a few extra allocations, or more code w/a few less allocations?

I suppose in an ideal world, ts-proto would let the user pick, but we already have a ton of flags, so dunno, I think I'd prefer to just pick one.

@sgielen wdyt about `|| []`?

Reply:
I'd not worry about performance too much. JIT engines run this stuff amazingly fast these days. The better the optimizer understands the intend of the code, the better.

I'd use `Object.entries(message.foo ?? {})` to
1. Only use the fallback in case of `null` and `undefined` and not bugs like `""`, `0` or `false`,
2. Always have an onject as input to `Object.entries(`

Reply:
But this extra undefined check is only needed in case of `forceOptionals = true`? So maybe the code output should only be generated in that case. Same in the diff nelow.

Reply:
I can switch to `|| []` and `|| {}` no problem.

I'll respond to the "extra undefined check" comment in another thread on this PR. :)

Reply:
Although - if we switch to `|| []`, we do put empty lists on the wire, while if we keep `if (.. !== undefined)`, we don't put the list on the wire. I have a slight preference not to put empty lists on the wire. What do you both think?

Reply:
Derp, nevermind -- an empty map doesn't actually cause any writer calls, so there's no on-wire difference between an empty map and not writing the map at all. So, this is good to go.